### PR TITLE
Allow delegation to ancestor for schema definition

### PR DESCRIPTION
**Pull request by AlessandroMinali**
https://github.com/nesaulov/surrealist/issues/11

Allow explicit delegation to ancestor to define schema.

### CODE REVIEW

**Comment by nesaulov** [comment]
@AlessandroMinali looks alright to me, but could you please add some more specs on this feature (like specs on exception raising, multiple inheritance (mixins), delegating serialization to a class that is not a parent of serializable class) ?

**Comment by AlessandroMinali** [comment]
I added more test. Please confirm this is behaviour you want though:

```
class Host
  include Surrealist

  json_schema do
    { name: String }
  end

  def name
    'Parent'
  end
end

class Guest < Host
  delegate_surrealization_to Host

  def name
    'Child'
  end

  # expecting: { name: 'Child' }
end

class FriendOfGuest < Guest
  def name
    'Friend'
  end

  # expecting: { name: 'Friend' }
end

class Invite < Host; end

class InvitedGuest < Invite
  delegate_surrealization_to Host

  def name
    'Invited'
  end

  # expecting: { name: 'Invited' }
end
```

**Comment by nesaulov** [comment]
I find this part confusing:
```ruby
class Guest < Host
  delegate_surrealization_to Host

  def name
    'Child'
  end
  # expecting: { name: 'Child' }
end

class FriendOfGuest < Guest
  def name
    'Friend'
  end
  # expecting: { name: 'Friend' }
end
```
I think that delegation should not happen implicitly (without explicit `delegate_surrealization_to`) , so `FriendOfGuest.new.surrealize` should raise an error (`Surrealist::UnknownSchemaError` for example).
So this part
```ruby
class << self
  def inherited(subclass)
    subclass.instance_variable_set('@__surrealist_schema_parent', @__surrealist_schema_parent)
  end
end
``` 
will not be necessary too.

Also, could you please write specs for a case where classes are not coupled at all, like this:
```ruby
class RandomClass
  include Surrealist

  json_schema do
    { name: String }
  end
end

class DifferentClass
  include Surrealist

  delegate_surrealization_to RandomClass

  def name
    'smth'
  end

  # expecting: { name: 'smth' }
end
```
This should also work.

**Comment by AlessandroMinali** [comment]
added that test case and made it pass. Please note these other tests cases you may have overlooked, just want to confirm you are getting the behaviour you want :)

```
# both are expected to fail
class InvalidGuest < Host
  delegate_surrealization_to 'InvalidHost'
end

class IncorrectGuest < Host
  delegate_surrealization_to Integer
end
```

File: lib/surrealist/class_methods.rb [hunk 1]
```diff
@@ -57,5 +57,41 @@ module ClassMethods
...
+    #       'Child'
+    #     end
+    #   end
+    #
+    #   Guest.new.surrealize
+    #   # => "{\"name\":\"Child\"}"
+    #   # For more examples see README
+    def delegate_surrealization_to(klass)
+      raise TypeError, "Expected type of Class got #{klass.class} instead" unless klass.is_a?(Class)
+      raise InvalidSchemaDelegation, 'Class does not include Surrealist' unless klass.included_modules.include?(Surrealist)
```
Comment (nesaulov) [inline lib/surrealist/class_methods.rb]: this line is too long, I think, maybe we should extract it to a method?
like `raise_invalid_delegation! unless ...`
or maybe convert modifier to statement

File: lib/surrealist/class_methods.rb [hunk 2]
```diff
@@ -57,5 +57,41 @@ module ClassMethods
...
+
+      self.instance_variable_set('@__surrealist_schema_parent', klass)
```
Comment (nesaulov) [inline lib/surrealist/class_methods.rb]: `self` is redundant here

**Comment by nesaulov** [comment]
@AlessandroMinali yeah, that's exactly what I've expected!
A couple last notes:
Could you please add spec for this example:
```ruby
class Potato < Vegetable
  delegate_surrealization_to Host

  def name
    'Potato'
  end
  # expecting: { name: 'Child' }
end

class Chips < Potato
  def name
    'Lays'
  end
  # expecting: Surrealist::UnknownSchemaError
end

context 'when parent class includes surrealist, but delegation is not specified' do
  it { expect { Chips.new.surrealize }.to raise_error(Surrealist::UnknownSchemaError, message) }
end
```
So that we are sure what error is raised here.
And also I've added some comments on the code

**Pull request merged**